### PR TITLE
dma: support for sys_dma_get_stream_info syscall

### PR DESCRIPTION
**Pull request by pthierry-ledger**
- [x] create a dedicated uapi dma.h header to properly separate DMA-related exported types
- [x] preparation: export informative-relative DMA types to uapi
- [x] add sysgate for sys_dma_get_stream_info
- [x] add autotest tests for this syscall
- [x] add Rust-related structure definition that match dma.h using public submodule in systypes module

This API return a complete structured stream definition based on device-tree. This avoid any device-tee manipulation at userspace level for DMAs.

Note: the overall DMA UAPI will have its associated full documentation (RustDoc+Concept/mans) through a dedicated, final, PR.

Note: some fields are bitfields (ORed values) making the UAPI in Rust harder to manipulate without depending on the public bitfield crate. The mask may need to be replaced by a full enumeration of cases.  This support is out of this very PR. See #68 for more information.

### CODE REVIEW

File: kernel/src/syscalls/meson.build [hunk 1]
```diff
@@ -28,6 +28,7 @@ syscalls = files(
     'sysgate_get_dmahandle.c',
     'sysgate_shm_set_creds.c',
     'sysgate_shm_get_infos.c',
+    'sysgate_dma_getinfo.c',
```
Comment (fvalette-ledger) [inline kernel/src/syscalls/meson.build:31]: `shm_get_infos` and `dma_getinfo` --> we should stay consistent in naming.

**Review by fvalette-ledger — APPROVED** [review]

File: kernel/src/managers/dma/dma-dt.c.in [hunk 2]
```diff
@@ -86,7 +86,7 @@ static const dma_meta_t streams[] = {
             .source = {{ stream_get_transfer_source(node) }},
             .dest = {{ stream_get_transfer_dest(node) }},
             {% if node["length"] -%}
-            .transfer_len = {{ "%08u"|format(node["length"]) }}UL,
+            .transfer_len = {{ "%u"|format(node["length"]) }}UL,
             {% else -%}
             .transfer_len = 0UL,
```
Comment (iartemov-ledger) [inline kernel/src/managers/dma/dma-dt.c.in:91]: Should we really accept the absence of `streamid` or `length` nodes and silently convert their values to 0 ?

File: kernel/src/managers/dma/dma-dt.c.in [hunk 3]
```diff
@@ -86,7 +86,7 @@ static const dma_meta_t streams[] = {
             .source = {{ stream_get_transfer_source(node) }},
             .dest = {{ stream_get_transfer_dest(node) }},
             {% if node["length"] -%}
-            .transfer_len = {{ "%08u"|format(node["length"]) }}UL,
+            .transfer_len = {{ "%u"|format(node["length"]) }}UL,
```
Comment (iartemov-ledger) [inline kernel/src/managers/dma/dma-dt.c.in:89]: Finally is not it too rigid to hard-code the transfer length in the DTS without possibility to adjust it with an API parameter of `sys_dma_start_stream()` par example? 
We might have use cases when we want to transfer different lengths up to a certain maxim (the one we could put in the DTS), might not we ?

File: uapi/include/uapi/dma.h [hunk 4]
```diff
@@ -0,0 +1,166 @@
+// SPDX-FileCopyrightText: 2023 Ledger SAS
```
Comment (iartemov-ledger) [inline uapi/include/uapi/dma.h:1]: 2024 ? ;)

File: uapi/include/uapi/dma.h [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+ * @enum gpdma_priority
+ *
+ * @brief DMA stream priority in comparison with others
+ */
+typedef enum gpdma_priority {
+    GPDMA_PRIORITY_LOW    = 0,
+    GPDMA_PRIORITY_MEDIUM = 1,
+    GPDMA_PRIORITY_HIGH = 2,
+    GPDMA_PRIORITY_VERY_HIGH = 3,
+} gpdma_priority_t;
```
Comment (iartemov-ledger) [inline uapi/include/uapi/dma.h:105]: Some of above structures are not used so far, correct?

Comment (pthierry-ledger) [reply uapi/include/uapi/dma.h:105]: prio is a dts entry, yet used at driver level, this is only canonical here

File: uapi/include/uapi/dma.h (hunk 4 above)
Comment (pthierry-ledger) [reply uapi/include/uapi/dma.h:1]: fixed

File: kernel/src/managers/dma/dma-dt.c.in (hunk 3 above)
Comment (pthierry-ledger) [reply kernel/src/managers/dma/dma-dt.c.in:89]: Yes, length is something specified at dts level by now.

Comment (pthierry-ledger) [reply kernel/src/managers/dma/dma-dt.c.in:89]: this is a protection mechanism to ensure that we can't dma-copy more len that what the shm support, at build time directly

File: kernel/src/managers/dma/dma-dt.c.in (hunk 2 above)
Comment (pthierry-ledger) [reply kernel/src/managers/dma/dma-dt.c.in:91]: yep. streamid is useful for mem2dev or dev2mem as it defines the HW stream config id that matches the corresponding dev flow. This is useless for mem2mem (detected by the fact that both src and dest are shm). in that last case, streamid as no meaning

**Pull request merged**